### ISSUE TITLE
FurAffinity plugin can't find submission ID if the URL has a comment ID in it

### ISSUE THREAD

**Issue by Eupeptic1** [Simon-Tesla/RaccoonyWebEx#83]
**Description**
If you visit a URL to a submission on FurAffinity that has a comment ID appended, Raccoony runs as normal and offers the "Download" button.  If you click the "Download" button, the file downloads, but it's missing the submission ID number at the beginning of the file name.  You expect something like `12345_my_nifty_art_by_joecool.png`, but you get `_my_nifty_art_by_joecool.png`.

**Steps to reproduce**
1.  Find any image on FurAffinity that has comments.  Example, sfw: https://www.furaffinity.net/view/40428437

2.  Pick one of the comments and click the '#link' button in the top right corner of the comment.  The URL will change to the direct link to that particular comment.  Example, sfw: https://www.furaffinity.net/view/40428437/#cid:153473681

3.  Mouse over the raccoon head and click the "Download" button.

4.  On the local filesystem, navigate to the directory where the file was saved.  Example: `~/Downloads/raccoony/furaffinity/may825`

5.  Look at the files in that directory.

**Expected result:**
There is a file for that submission, with a filename starting with FA's submission ID.  Example: `40428437_ki_taste_new_by_may825.jpg`
 
**Actual result:**
The file exists, but lacks the submission ID at the beginning.  Example: `_ki_taste_new_by_may825.jpg`

**Environment**
 Raccoony 1.2.3, Firefox 78.7.0esr, 64-bit, desktop, on Debian Linux 9.9, x86_64.

**Comment by Eupeptic1** [comment]
**Analysis**

The plugin runs as normal on a URL with a comment ID, but it has `submissionId: null` when it returns the `I.Media` struct to the caller.

`getMediaUrls()` seems to be doing the right thing, as far as parsing the direct link to the image out of the Download button on FA.

The flaw in the slaw seems to be in `getIdFromSubmissionUrl()`.  It currently does
`let match = url.match(/\/(\d+)\/?$/);
return match && match[1];`

and I think that regex is wrong.  It's looking for a forward slash, followed by one or more digits, followed by another forward slash, a (spurious) literal question mark OR optional match, followed by end of line.

I also don't quite understand the return statement.  I think it's trying to say "if the match worked, return the second element", but I'm not sure it's working.  If the regex is doing what I think, the match shouldn't work at all, and just `match` should be false/zero/null.

**Potential fix**
Make `getIdFromSubmissionUrl()` smarter.  The obvious thing is to trim `#.*$` off the end of the URL before doing anything else.  That might not be the right thing in all cases, though.

**Comment by Eupeptic1** [comment]
I didn't notice there had been a release.  :D  I updated to 1.2.4 and I can still reproduce the problem.

**Comment by Eupeptic1** [comment]
**More analysis**

The regex in `getIdFromSubmissionUrl()` is definitely the problem.  I have modified it to avoid this bug, but I'm still playing with it.

Something possibly interesting: if I start with the current code, and only add two logging lines right before the return in that function:

`logger.log("fa: match ", match);`

`logger.log("fa: match1 ", match[1]);`

then when I load a page with a comment ID on the URL, I see `fa: match null` in the console, the second line doesn't print in the console, _and I never get the raccoon head on the page_.  I'm assuming that since `match[1]` doesn't exist, trying to log it makes something explode, so TypeScript treats that function as having returned null, zero, bad, nothing, which makes the rest of Raccoony think there was an error in parsing.

In other words, the current `return match && match[1]` isn't catching the case where `match` is null.  Maybe it would work if `match` was totally undefined.

Also, apparently, `match` returns the whole substring matched (like `/40428437/` for my modified regex), while `match[1]` gives me just the bit in parentheses in the the regex (like `40428437`).  So I guess the 1 index is like using \1 in sed, vi, or equal.

**Comment by Eupeptic1** [comment]
Yeah, the current `return` statement isn't DTRT.  If I modify the regex so that it never will match, and replace the current `return` with

    if(match) {
        return(match[1]);
    } else {
        return("unknown");
    }

then it returns "unknown" to the caller, I get the raccoon head, and the file gets saved on local disk with "unknown" as the first part of the filename.

I don't know how _useful_ this information is, but I figured I'd note it.  I guess if there are other plugins that use that "thing-that-shouldnt-be-null && thing-that-shouldnt-be-null[1]" idiom, it might be a good idea to review those statements.

**Comment by Eupeptic1** [comment]
Possible fix now on my fork.

I changed the regex to look for `/view/[numbers]/` in the URL.  If the regex doesn't match, it returns the text string "unknown".  This is changed behavior, but maybe it is better than just returning an empty string.

**Comment by Simon-Tesla** [comment]
Ah! Thank you for looking into this n.n

**Comment by Eupeptic1** [comment]
I am running my modified version now.  I'm going to test it on more submission types (music, animations, etc) and generally dogfood it for a few days before doing anything else.

**Comment by Eupeptic1** [comment]
One thing I have noticed in testing: if you turn the metadata file on, and download from a link with a comment ID in it, then the comment ID is still in the "Source URL" fields in both the plain text and JSON.

Is this important?  I can probably fix it, but I wanted to get your opinion.

The case I can think of where it *might* cause a problem:

- user has metadata files turned on
- user clicks on a comment link on FA
- user downloads the image with Raccoony
- either the person who made that comment, or the artist, or an admin hides the comment on FA
- user decides that they'd like to visit that submission page again
- user looks up URL in metadata file and tries to visit it
- FA 404s or otherwise can't find the submission

I guess the key here is "what happens if you try to follow a link to a hidden comment on FA?"  If FA is smart enough to chop off the comment ID and just present the submission page, then it's not a big deal.  If FA 404s or otherwise errors out, then probably I should make source URL be "canonical" - without the comment ID.

The other URL fields in the JSON (`url`, which is the direct link to the image, and `previewUrl`, the thumbnail), are still set correctly.  The plain text doesn't have any other URL fields.

I've tested this on many still images and it appears to be working OK.  I still need to test it on some music and animations before I send the pull request.

**Comment by Eupeptic1** [comment]
The new regex that fixes this issue is independent of whether the URL is normalized or not.  So for now I will plan on not normalizing the URL.

**Comment by Eupeptic1** [comment]
**DON'T SHIP THIS, IT STILL HAS A BUG.**

More soon.

Not that you're up cutting releases at 3 AM, though.  :D

**Comment by Eupeptic1** [comment]
The regex I changed doesn't consider the case where the submission ID is the very last thing in the URL, and there is no trailing slash on the URL.  It can't find the submission ID in that case and sets it to "unknown".

#### Steps to reproduce:

1. Load up Raccoony and open the web console.
2. Go to a URL on FA, like https://www.furaffinity.net/view/30529254 (sfw).  *Do not include the trailing slash in the URL.*
3. Look in the web console for this line `[raccoony] sending message Object { action: "messageaction_checkdownload", data: {…} } page_inject.js:32332:20` , expand the top-level `Object`, and then expand the `data` inside that.
4. You should expect to see a valid FA submission ID listed under `submissionId`, but you will probably see `"unknown"` instead.
5. (Optional) Click the Raccoony Download button.  The file will save in the correct directory of `~/Downloads/raccoony/furaffnity/[artist]/` .  You should expect the filename to start with the FA submission ID, but it will probably start with `unknown_`.
5. In your browser address bar, *add the trailing slash to the URL*, and hit Enter to reload the page.
6. Look in the web console again, and expand the same line.
7. You will probably see a valid FA submission ID listed under `submissionId`.
8. (Optional) Hit **D** to force a re-download.  The file will save in the correct directory of `~/Downloads/raccoony/furaffnity/[artist]/` .  The filename will start with the FA submission ID, as expected.

I didn't see this until now because Raccoony's open-all-in-tabs function includes the trailing slash in the URL for each new tab it opens.  When I just did a straight "download" with Raccoony on that page, the regex was happy.  If I clicked on one of the comment links on the submission to put the comment ID in the URL (like https://www.furaffinity.net/view/30529254/#cid:136386771 ), the comment ID was appended after the trailing slash, so if I then told Raccoony to download, the regex was still happy.

I found it when I was doing my semi-automated testing for #90.  The file of test bookmarks I use does not have the trailing slashes on most of the URLs, and each bookmark gets opened by another plugin, so it never gets the trailing slash put on it.  I did the test in three browsers and they all did exactly the same (wrong) thing, but I didn't notice.  I realized I didn't have any "truth" to compare my changes to, so I loaded up vanilla Raccoony 1.2.4 and re-ran the test in Firefox.  I got the same *quantity* of files, and the checksums for the corresponding files matched, but the file names did not.

The solution is to fix the regular expression.  (Somewhere in a nightclub in the Bay Area, jwz is laughing at me.)

**Comment by Eupeptic1** [comment]
The original regex, before I changed anything for this issue, was `/\/(\d+)\/?$/` .  This allowed for the optional slash at the end of the URL, but enforced that either the submission ID or the slash was at the end of the URL, which made it not find the submission ID if the URL had a comment ID.

The modified regex in the fix and pull request above, which turned out to be incorrect, was `/\/view\/(\d+)\//` .  This enforced a slash appearing after the submission ID, which is not always the case; https://www.furaffinity.net/view/40428437 works just as well as https://www.furaffinity.net/view/40428437/ .

I have further modified the regex to `/\/view\/(\d+)/`, which only enforces `/view/`, followed by a number of digits.  This appears to work better in automated testing and limited dogfooding.  I want to dogfood it some more before I declare it fixed.

**Comment by Simon-Tesla** [comment]
Sounds good. I'm looking at seeing if I can put together a release this weekend, but I'll hold off until you've got a fix ready.

**Comment by Eupeptic1** [comment]
I don't want to block a release.  I uploaded the fix for this and for #90 to my master, so you can test if you want.  I haven't created a PR yet, though.

(I feel a little silly that I dogfooded my previous "fix" for a couple of weeks and didn't catch this, so I'm a little shy of saying "for sure this time, lol".  I also note that the original regex, before I changed anything, already accounted for the possibility of the trailing slash being there or not, but I didn't completely notice what it was doing.)

**Comment by Simon-Tesla** [comment]
Eh, it's not a huge deal, I can hold off. Might work on a few other fixes in the meantime, and the original fix isn't something that must go out now either.

**Comment by Eupeptic1** [comment]
I am still dogfooding this.  I have downloaded about 75 submissions so far, with various combinations of slash / no slash / comment ID / no comment ID, and it appears to work OK.  When I get to 100, I'll create a PR for it.  This should happen some time in the next few days.

**Comment by Eupeptic1** [comment]
I found an alt of an artist I was already following on FA, so I was able to download more new submissions than I usually can.  That got me to over 112 downloads with this fix, and they all appeared to work correctly.

I had 28 downloads each for the following FA URL formats.  For all of them, the FA plugin found the submission ID, named the downloaded file starting with the submission ID, and placed the submission ID in the metadata file.

    /view/12345#cid:6789     (no trailing slash, comment ID)
    /view/12345/#cid:6789    (trailing slash,    comment ID)
    /view/12345              (no trailing slash, no comment ID)
    /view/12345/             (trailing slash,    no comment ID)
